File: src/python/problems/inorder_traversal_test.py
# https://leetcode.com/problems/binary-tree-inorder-traversal

from typing import Optional
from core.problem_base import *
from models.tree_node import *

class InorderTraversal(ProblemBase):
    def Solution1(self, root: Optional[TreeNode]) -> list[int]:
        result = []

        def Dfs(node):
            if not node:
                return
            
            Dfs(node.left)

            result.append(node.val)

            Dfs(node.right)

        Dfs(root)

        return result
    
    def Solution2(self, root: Optional[TreeNode]) -> list[int]:
        result = []
        stack = []

        node = root
        while node or stack:
            if node:
                stack.append(node)
                node = node.left
            else:
                pop = stack.pop()
                result.append(pop.val)
                node = pop.right

        return result

if __name__ == '__main__':
    TestGen(InorderTraversal) \
        .Add(lambda tc: tc.ParamTreeNode([1,None,2,3]).Result([1,3,2])) \
        .Add(lambda tc: tc.ParamTreeNode([]).Result([])) \
        .Add(lambda tc: tc.ParamTreeNode([1]).Result([1])) \
        .Run()

File: src/challenges/go/problems/inorder_traversal_test.go
// https://leetcode.com/problems/binary-tree-inorder-traversal

package problems

import (
	"testing"
	"github.com/dobermanch/leetcode/core"
)

type InorderTraversal struct{}

func TestInorderTraversal(t *testing.T) {
	gen := core.TestSuite[InorderTraversal]{}
	gen.Add(func(tc *core.TestCase) {
		node := &TreeNode{ Val: 1, Right: &TreeNode{ Val: 2, Left: &TreeNode{Val: 3}}}

		tc.Param(node).Result([]int{1,3,2})
	}).Run(t)
}

func (InorderTraversal) Solution1(root *TreeNode) []int {
	result := []int{}

	var dfs func(node *TreeNode)
	dfs = func(node *TreeNode) {
		if node == nil {
			return
		}
		
		dfs(node.Left)

		result = append(result, node.Val)

		dfs(node.Right)
	}

	dfs(root)

	return result
}

func (InorderTraversal) Solution2(root *TreeNode) []int {
	result := []int{}
	stack := []*TreeNode{}

	node := root
	for node != nil || len(stack) > 0 {
		if node != nil {
			stack = append(stack, node)
			node = node.Left
		} else {
			var pop = stack[len(stack) - 1]		
			result = append(result, pop.Val)
			node = pop.Right
			stack = stack[:len(stack) - 1]
		}
	}

	return result
}
